File: Practica/src/Guias/Guia3/Ej7/BankAccount.java
package Guias.Guia3.Ej7;

public abstract class BankAccount {

    private int id;
    private double balance;

    public BankAccount(int id) {
        this.id = id;
        this.balance = 0;
    }

    public void deposit(double amount) {
        this.balance += amount;
    }

    public void extract(double amount) {
        this.balance -= amount;
    }

    protected abstract boolean canExtract(double amount);

    public double getBalance() {
        return this.balance;
    }

    @Override
    public String toString() {
        return "Cuenta %d con saldo %.2f".formatted(id, balance);
    }

}

File: RUBY/Guias/Guia8/Ej5_bank_account/bank_account.rb


class BankAccount
  def initialize
    raise "Debe sobreescribir los métodos"
  end
  
  def create(id)
    @id = id
    @balance = 0
  end

  def deposit(amount)
    @balance =+ amount
  end

  def extract(amount)
    raise "No cuenta con el balance necesario" unless extract? amount
    @balance -= amount
  end

  def extract?(amount)
    raise "Debe sobreescribir el método"
  end

  def to_s
    "Cuenta #{@id} con saldo #{@balance}"
  end

  private :create, :extract?

end

class CheckingAccount < BankAccount
  def initialize(id, overdraft)
    create(id)
    @overdraft = overdraft
  end

  private def extract?(amount)
    @balance + @overdraft - amount >= 0
  end
end

class SavingsAccount < BankAccount
  def initialize(id)
    create(id)
  end

  private def extract?(amount)
    @balance - amount >= 0
  end
end





